src/prefect/engine.py

```diff
@@ -2056,6 +2065,32 @@ def should_log_prints(flow_or_task: Union[Flow, Task]) -> bool:
     return flow_or_task.log_prints
 
 
+def _resolve_custom_flow_run_name(flow: Flow, parameters: Dict[str, Any]) -> str:
+    if callable(flow.flow_run_name):
+        flow_run_name = flow.flow_run_name()
+        if not isinstance(flow_run_name, str):
+            raise RuntimeError("callable 'flow_run_name' did not return a string")
```

Comment:
Note all of these `RuntimeError`s can be `TypeError`s

Reply:
Hey, I'll convert `RuntimeError`s to `TypeError`s where we check the type of the generating function/string. I'm not sure `TypeError` should be raised for the return value of the function.

Reply:
Just checked that `ValueError` is being raised on invalid return values in the code-base. Since the practice is established, I'll raise `TypeError` for the non string return value.

Reply:
done